File: dlmaim.m
function q = dlmaim(R3, O3, A3, S)
% DLMAIM
% Detail Loss Measure with Additive Impairment Modulation.
%
% Inputs:
%   R3 - absolute value of restored/detail coefficients
%   O3 - absolute value of reference coefficients
%   A3 - absolute value of additive impairment coefficients
%   S  - bookkeeping matrix from wavedec2
%
% Output:
%   q  - scalar quality score
%
% NOTE:
%   This implementation follows the ADM-style public code used in the
%   literature (e.g., Li et al., IEEE). It operates directly on the
%   wavedec2 coefficient vector using explicit subband indexing.

% -------------------------
% Part 1: Detail restoration term
% -------------------------
nume = 0;
deno = 0;

for lambda = 1:4
    for theta = 2:4   % detail subbands only

        [s, e] = getSubbandSE(lambda, theta, S);

        Rsub = reshape(R3(s:e), S(6-lambda,1), S(6-lambda,2));
        Osub = reshape(O3(s:e), S(6-lambda,1), S(6-lambda,2));

        % Remove border region (center crop)
        Rsub = centerCrop(Rsub);
        Osub = centerCrop(Osub);

        a = sum(Rsub(:).^3)^(1/3);
        b = sum(Osub(:).^3)^(1/3);

        nume = nume + a;
        deno = deno + b;
    end
end

q1 = nume / (deno + eps);

% -------------------------
% Part 2: Additive impairment penalty
% -------------------------
nume = 0;
deno = numel(A3);

for lambda = 1:4
    for theta = 2:4

        [s, e] = getSubbandSE(lambda, theta, S);

        Asub = reshape(A3(s:e), S(6-lambda,1), S(6-lambda,2));
        Asub = centerCrop(Asub);

        SD = std(Asub(:));
        a = sum((Asub(:) + SD).^3)^(1/3);

        nume = nume + a;
    end
end

a1 = 10;   % tuning constant (kept from original code)
q2 = nume / (deno + eps);

% -------------------------
% Final DLMAIM score
% -------------------------
q = q1 - a1 * q2;

end

% ==========================================================
% Helper functions
% ==========================================================

function B = centerCrop(A)
% CENTERCROP Remove border region to avoid boundary artifacts
marg = 0.1;  % 10% margin
[h, w] = size(A);
mh = max(1, round(h * marg));
mw = max(1, round(w * marg));
B = A(mh:h-mh+1, mw:w-mw+1);
end

function [s, e] = getSubbandSE(lambda, theta, S)
% GETSUBBANDSE Return start/end indices of a subband
if lambda == 4 && theta == 1
    s = 1;
    e = S(1,1) * S(1,2);
    return;
end

lambda2 = 5 - lambda;
s = S(1,1) * S(1,2) + 1;

for i = 2:lambda2
    s = s + 3 * S(i,1) * S(i,2);
end

theta2 = theta - 2;
s = s + theta2 * S(lambda2+1,1) * S(lambda2+1,2);
e = s + S(lambda2+1,1) * S(lambda2+1,2) - 1;
end

File: python/dlmaim.py
import numpy as np


def _center_crop(A, marg=0.1):
    """Remove a 10% border to avoid boundary artifacts (mirrors MATLAB centerCrop)."""
    h, w = A.shape
    mh = max(1, round(h * marg))
    mw = max(1, round(w * marg))
    # MATLAB A(mh:h-mh+1, ...) is 1-indexed → Python A[mh-1:h-mh+1, ...]
    return A[mh - 1:h - mh + 1, mw - 1:w - mw + 1]


def dlmaim(R_coeffs, O_coeffs, A_coeffs):
    """
    Detail Loss Measure with Additive Impairment Modulation.

    Parameters
    ----------
    R_coeffs : list — abs of restored coefficients,  pywt.wavedec2 structure
    O_coeffs : list — abs of reference coefficients, pywt.wavedec2 structure
    A_coeffs : list — abs of additive-impairment coefficients, same structure

    Each list: [approx_arr, (H_n, V_n, D_n), ..., (H_1, V_1, D_1)]

    Returns
    -------
    q : float — DLMAIM quality score
    """
    eps = np.finfo(np.float64).eps

    # ------------------------------------------------------------------
    # Part 1: detail restoration term  (L3-norm ratio across all subbands)
    # ------------------------------------------------------------------
    nume1 = 0.0
    deno1 = 0.0
    for i in range(1, len(R_coeffs)):        # skip approximation band (index 0)
        for theta in range(3):               # H=0, V=1, D=2
            Rsub = _center_crop(R_coeffs[i][theta])
            Osub = _center_crop(O_coeffs[i][theta])
            nume1 += float(np.sum(Rsub.ravel() ** 3) ** (1.0 / 3.0))
            deno1 += float(np.sum(Osub.ravel() ** 3) ** (1.0 / 3.0))

    q1 = nume1 / (deno1 + eps)

    # ------------------------------------------------------------------
    # Part 2: additive impairment penalty
    # ------------------------------------------------------------------
    # Total coefficient count (approximation + all detail subbands)
    total_size = A_coeffs[0].size + sum(
        A_coeffs[i][theta].size
        for i in range(1, len(A_coeffs))
        for theta in range(3)
    )

    nume2 = 0.0
    for i in range(1, len(A_coeffs)):
        for theta in range(3):
            Asub = _center_crop(A_coeffs[i][theta])
            SD = float(np.std(Asub))
            nume2 += float(np.sum((Asub.ravel() + SD) ** 3) ** (1.0 / 3.0))

    a1 = 10.0
    q2 = nume2 / (total_size + eps)

    return q1 - a1 * q2
